### DISCUSSION TITLE
Changing keyboard layout in each window independently

### DISCUSSION THREAD

**Discussion by mohamadali-halwani** [i3/i3#4916] [Ideas]
Hello there [:)](https://tenor.com/bzu1H.gif)

I switched from KDE a while ago, and this is what I miss the most from KDE.

I would like that when I switch the keyboard layout in one window, the other window's keyboard layout stays the same. The reason why I want this is because mostly I chat and write code simultaneously, and it's a pain to keep switching the layout each time I switch the window.
Here is the option I meant

```
Switching Policy

If you select Application or Window switching policy, changing the keyboard layout will only affect the current application or window. 
Global will change the keyboard layout globally and Desktop will change it only for the current virtual desktop.

```
https://docs.kde.org/stable5/en/plasma-desktop/kcontrol/keyboard/layouts.html

I tried deploying the script here https://www.reddit.com/r/i3wm/comments/9flipu/comment/e60n5ah/?utm_source=share&utm_medium=web2x&context=3 and it did nothing.

Thanks in advance.

**Comment by mohamadali-halwani** [comment]
Someone pointed me to the specific code
**In case of Wayland you can find the relevant code here:** https://invent.kde.org/plasma/kwin/-/blob/master/src/keyboard_layout_switching.cpp#L201

```
WindowPolicy::WindowPolicy(KWin::Xkb* xkb, KWin::KeyboardLayout* layout)
    : Policy(xkb, layout)
{
    connect(workspace(), &Workspace::clientActivated, this,
        [this] (AbstractClient *c) {
            if (!c) {
                return;
            }
            // ignore some special types
            if (c->isDesktop() || c->isDock()) {
                return;
            }
            setLayout(getLayout(m_layouts, c));
        }
    );
}

WindowPolicy::~WindowPolicy()
{
}

void WindowPolicy::clearCache()
{
    m_layouts.clear();
}

void WindowPolicy::layoutChanged(uint index)
{
    auto c = workspace()->activeClient();
    if (!c) {
        return;
    }
    // ignore some special types
    if (c->isDesktop() || c->isDock()) {
        return;
    }

    auto it = m_layouts.find(c);
    if (it == m_layouts.end()) {
        m_layouts.insert(c, index);
        connect(c, &AbstractClient::windowClosed, this,
            [this, c] {
                m_layouts.remove(c);
            }
        );
    } else {
        if (it.value() == index) {
            return;
        }
        it.value() = index;
    }
}

```

**Comment by orestisfl** [comment]
i3 does not know about keyboard layouts so your best bet would be to get that script working. Does it log any error?

**Reply by mohamadali-halwani** [reply]
At least if i3 can recognize an event like focusing on a window, that would make things easier. maybe using a script combining `i3` and `setxkbmap`.
That python script did not show any errors.

**Reply by orestisfl** [reply]
Yes, i3 recognizes and provides focus events. The linked script uses them.
